### PR TITLE
add string context 255

### PR DESCRIPTION
**Pull request by gbwey9**
Hi, 

Here is my attempt at doing #255 to add string context. 

I am particularly unsure about this change that I made in Eval.hs in MonadEval  
-    evalString      :: NString (m v) -> m v
+    evalString      :: NixString -> m v

Anyway let me know what I need to fix.

Best,
Grant

### CODE REVIEW

**Comment by shlevy** [comment]
Will take a deep look this weekend!

File: src/Nix/Builtins.hs
Comment (jwiegley) [inline src/Nix/Builtins.hs]: I really like using `NixString` to abstract this tuple, and now can't believe it hadn't been done until now. :)

File: src/Nix/Builtins.hs
Comment (jwiegley) [inline src/Nix/Builtins.hs]: I'd prefer if the qualified name was just `S`, and so didn't encode anything about which *kind* of Set is being used, since we might change that in future.

File: src/Nix/Eval.hs
Comment (jwiegley) [inline src/Nix/Eval.hs]: The `:: String` qualification isn't needed here, because `String` is the only type that `ErrorCall` accepts.

File: src/Nix/Value.hs
Comment (jwiegley) [inline src/Nix/Value.hs]: Good stuff, thought I'll have to defer to @shlevy as to its correctness.

File: src/Nix/Value.hs
Comment (shlevy) [inline src/Nix/Value.hs]: `DerivationOutput` should take a `Text` (referring to the specific output requested, e.g. `"out"`)

File: src/Nix/Value.hs
Comment (shlevy) [inline src/Nix/Value.hs]: Nit: `AllDerivationOutputs`

**Review by shlevy — CHANGES_REQUESTED** [review]
I'm thinking we should move the definition of `NixString` to its own module and not export the constructor/field names. We really don't want things like the implicit dropping of context in the `fromValueMay` definition, for example.

Instead, we should have something like (names certainly subject to change) `stringNoContext` which converts a `NixString` to a `Maybe Text`, being `Nothing` if the `NixString` has context, `stringIntentionallyDropContext` which converts a `NixString` to a `Text`, and `stringWithContext` which gives you both the `Text` and the `HashSet StringContext`. We should make it hard to use a `NixString` without explicitly thinking about what we want to do with the context.

**Comment by jwiegley** [comment]
@shlevy I very much agree on moving it to its own module. We need more abstraction of that sort in `hnix`.

File: src/Nix/Builtins.hs
Comment (mightybyte) [reply src/Nix/Builtins.hs]: Yeah, this is vastly more understandable.

**Comment by gbwey9** [comment]
Yes, moving NixString to its own module is a great idea! I'll get on that and also the other changes suggested by John above.

**Comment by shlevy** [comment]
@gbwey9 Thanks! I'll take another pass this weekend.

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: We should probably have a separate error type for this use case (maybe even just make `stringNoContext` throw it?). This is the error message C++ Nix gives us:

> nix-repl> builtins.getAttr "${./.}"  {} 
> error: the string '/nix/store/a8bwd9g6dvhbcr5q6pgh84i80zc0zxfq-hnix-store' is not allowed to refer to a store path (such as '/nix/store/a8bwd9g6dvhbcr5q6pgh84i80zc0zxfq-hnix-store'), at (string):1:1

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs:471]: Wait, does this imply there's a `fromValue` instance that gives you `Text`? We definitely don't want that!

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: `pathExists` should fail if the string has context.

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: In the C++ code, `lessThan` only compares the underlying strings, not the contexts... what does `Ord` for `NixString` do?

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: For now, we should fail if the string has context, but actually in the C++ code base this works through the same mechanism as import-from-derivation: If we do readFile or readDir on a path with conext, we build the derivations we need to build during evaluation and *then* read from the file/directory.

@jwiegley What's the best way to track that we need to change this once we support IFD?

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: `toXML` should output a `NixString`, it may have some context depending on the value in question.

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: URL can't have any context.

File: src/Nix/Builtins.hs
Comment (shlevy) [inline src/Nix/Builtins.hs]: URL can't have any context.

File: src/Nix/Eval.hs
Comment (shlevy) [inline src/Nix/Eval.hs]: Hmm, why did this change?

File: src/Nix/Exec.hs
Comment (shlevy) [inline src/Nix/Exec.hs]: Hmm it's hard to tell here what `evalString` is doing... @jwiegley What's this used for? We definitely shouldn't drop context in a generic function...

File: src/Nix/Exec.hs
Comment (shlevy) [inline src/Nix/Exec.hs]: Not sure where `coerceToString` is being used, but if it's inspired by the function of the same name in the C++ codebase we definitely don't want to drop context here.

File: src/Nix/Normal.hs
Comment (shlevy) [inline src/Nix/Normal.hs]: We want to add the new path to the context here, not drop it.

File: src/Nix/Normal.hs
Comment (shlevy) [inline src/Nix/Normal.hs]: What is this function for?

File: src/Nix/XML.hs
Comment (shlevy) [inline src/Nix/XML.hs]: This ultimately needs to be changed to output a NixString, not a Text

**Review by shlevy — CHANGES_REQUESTED** [review]
Sorry for the large number of changes... Let me know if you want to chat synchronously at some point to help figure out the best way forward, or if you just want to keep going addressing one by one.

File: src/Nix/Builtins.hs
Comment (jwiegley) [reply src/Nix/Builtins.hs]: A separate error type would be nice.

File: src/Nix/Builtins.hs
Comment (jwiegley) [reply src/Nix/Builtins.hs:471]: Why not? It's OK to drop context info you're not going; you just don't want a `toValue` that forgets it when constructing a new string.

File: src/Nix/Builtins.hs
Comment (jwiegley) [reply src/Nix/Builtins.hs]: Good catch.

File: src/Nix/Builtins.hs
Comment (jwiegley) [reply src/Nix/Builtins.hs]: Track it with a GitHub issue?

File: src/Nix/Exec.hs
Comment (jwiegley) [reply src/Nix/Exec.hs]: `evalString` is recording provenance while building a string. Sounds like context needs to be preserved here.

File: src/Nix/Exec.hs
Comment (jwiegley) [reply src/Nix/Exec.hs]: It's meant to become an exact replica of the C++ version.

File: src/Nix/Builtins.hs
Comment (shlevy) [reply src/Nix/Builtins.hs:471]: @jwiegley I think we want to make it hard to accidentally drop context. Some primops should fail on non-empty context, others need to do special work (e.g. IFD) on non-empty context, even if there's no new string being constructed.

**Comment by shlevy** [comment]
@gbwey9 I was thinking it might be useful for me to write up an extensive document describing string context from the perspective of Nix language semantics, abstracted away from the specific implementation. Does that seem helpful?

**Comment by gbwey9** [comment]
@shlevy Yes a document describing string context would be really helpful, thanks! What would you recommend for getting up to speed on Nix (or at least the stuff I need to know)? I am currently going through the Nix pills series.

**Comment by shlevy** [comment]
@gbwey9 For the stuff you're working on, I definitely recommend Eelco's thesis! https://nixos.org/~eelco/pubs/phd-thesis.pdf

**Comment by gbwey9** [comment]
@shlevy Thanks I'll take a look.

**Comment by shlevy** [comment]
Hi @gbwey9! So sorry I dropped this! Are you still interested in working on it if I get that doc together for you?

**Comment by gbwey9** [comment]
Hi @shlevy , I am still interested in working on this and would be really happy to get that doc from you.
However, is there some way to break up this pull request into smaller more manageable pieces?
I am not sure if this is possible otherwise, as nix string impacts so much of the code.

**Comment by jwiegley** [comment]
@gbwey9 As a very first step, let's abstract out a `NixString` data type, and plumb that change throughout the code, and after that start adding context info.

**Comment by gbwey9** [comment]
@jwiegley Great idea. I'll work on that.

**Comment by shlevy** [comment]
OK, great! I'll pick up the documentation work then, and we can figure out the right path from there.

**Comment by gbwey9** [comment]
As you suggested, I have plumbed Nix through the codebase. 
Let me know what corrections I need to make.
Best,
Grant

**Comment by shlevy** [comment]
@gbwey9 A bunch of us are gathering this weekend (you're welcome to join if you're near NH or want to connect remotely!), we'll take another look at this then!

**Comment by gbwey9** [comment]
@shlevy that sounds great. I can't make it in person but would be glad to connect remotely. 
I do appreciate any help you and the others can give me on this

**Comment by jwiegley** [comment]
@gbwey9 Maybe we could phone you in by Skype?

**Comment by gbwey9** [comment]
@jwiegley Sounds good. However, can i just call in? I don't yet have a Skype account but I can set one up if that will work better.

**Comment by jwiegley** [comment]
@gbwey9 Yes, we could put you on speaker phone for a bit.

**Comment by gbwey9** [comment]
@jwiegley Thanks.

**Comment by shlevy** [comment]
@gbwey9 How should we contct you?

**Comment by gbwey** [comment]
my number is 1-734-274-5299 or I can call you.

**Comment by shlevy** [comment]
@gbwey texted you

**Comment by gbwey** [comment]
Sorry I have no idea where that text went. but my email address is weyburne@gmail.com

**Comment by jwiegley** [comment]
@gbwey @shlevy How is our progress on the string context stuff? Have we finished creating a new NixString type?

**Comment by gbwey9** [comment]
@jwiegley I have been actively working on it this past week and will ask for some more help from Shea in order to finish it off.

File: src/Nix/NixString.hs
Comment (shlevy) [inline src/Nix/NixString.hs]: We should have a `hacky` variant of this and use it everywhere for now. While making a string without context is something we want to expose, we still need to go through all current string creation to ensure it ought to be contextless

File: src/Nix/NixString.hs
Comment (shlevy) [inline src/Nix/NixString.hs]: This one is actually non-hacky for free :)

File: src/Nix/Builtins.hs
Comment (shlevy) [reply src/Nix/Builtins.hs:471]: @gbwey9 FYI, this is OK for now but once you try to remove the `hacky` bits from the `FromValue` instance we'll definitely have to address this properly.

File: src/Nix/NixString.hs
Comment (shlevy) [inline src/Nix/NixString.hs]: I think I'd prefer a `hacky` version of this as well, most of these are probably correct as-is but some string modifications absolutely need to change context as well.

File: src/Nix/NixString.hs
Comment (shlevy) [inline src/Nix/NixString.hs]: Can we have some more haddocks on these constructors?

File: src/Nix/NixString.hs
Comment (shlevy) [inline src/Nix/NixString.hs]: Let's leave this for now, but I am a tiny bit concerned that it might let context mishandling sneak in somewhere. Just to flag it in case we're seeing weird bugs in the future :joy:

**Review by shlevy — CHANGES_REQUESTED** [review]
Looking great, much nicer! In addition to inline comments, a few top-level thoughts:

1. It seems like your editor marked all files executable and changed newline endings for the READMEs/LICENSEs. I'm wondering if that's some Windows setting, can you see if that can be fixed?
2. The commit history is quite messy here, once this is stabilized we should probably squash it down to something more useful to people looking at the logs

File: src/Nix.hs
Comment (jwiegley) [inline src/Nix.hs]: I think, for consistency's sake, that while `NixString` is a fine name for the type, the module is better named just `Nix.String`. `Nix.NixString` seems redundant...

File: src/Nix/Builtins.hs
Comment (jwiegley) [inline src/Nix/Builtins.hs:104]: Does the function really need to be named `hackyMakeNixStringWithoutContext`? Is this some indication that the work is yet undone?

**Review by jwiegley — CHANGES_REQUESTED** [review]
It's starting to look pretty good!

File: src/Nix/Builtins.hs
Comment (shlevy) [reply src/Nix/Builtins.hs:104]: Yes. We need to ensure we look at all places where we create strings and set the context appropriately. For some of those it will be correct to just make a string without context, of course, but currently *every* string creation in hnix is without context and that's definitely wrong.

Comment (gbwey9) [reply src/Nix/Builtins.hs:104]: @shlevy As you mentioned before, there are two more places where we use mconcat (Eval.hs) and mappend (Exec.hs) on NixStrings. Should we just get rid of the semigroup/monoid instance and replace those calls with NixString specific calls?

File: src/Nix/Exec.hs
Comment (jwiegley) [inline src/Nix/Exec.hs:339]: Can you explain a bit more of the motivation behind this?

**Review by jwiegley — CHANGES_REQUESTED** [review]

File: src/Nix/Exec.hs
Comment (gbwey9) [reply src/Nix/Exec.hs:339]: It may be that some of the NixString function calls are incorrect, so the first step is to identify every instance. Then we can systematically evaluate each occurrence and then assign it a non-hacky* function from NixString. In other words, this first step is solely to mark these NixString calls.

File: src/Nix/Builtins.hs
Comment (shlevy) [reply src/Nix/Builtins.hs:104]: @gbwey9 Yeah, I think that makes the most sense. We can add it back in once we've cleaned everything up and found the right patterns for using context.

File: src/Nix/Exec.hs
Comment (jwiegley) [reply src/Nix/Exec.hs:339]: As long as at some point in the future we go back to using `Monoid` if this thing turns out to be a monoid, I'm fine with interim measures.

**Comment by gbwey9** [comment]
I just noticed that I missed the NixString change from Nix.NixString to Nix.String.
Also the travis build (strict) fails because of the deprecation warnings that are on the some of the NixString methods but builds otherwise. Should I remove those warnings from Nix.String?

**Comment by gbwey9** [comment]
@shlevy @jwiegley Is there anything else I need to do to get this pull request through?
I am happy to work on this, so just let me know.

**Review by shlevy — APPROVED** [review]
Looks great! I'd prefer a squash before merge but I'll leave that up to @jwiegley

**Comment by jwiegley** [comment]
I avoid squash, both at the dinner table and in Git. I'll merge once CI is happy.

**Review by jwiegley — APPROVED** [review]

**Comment by gbwey9** [comment]
@jwiegley @shlevy  Thanks so much

**Pull request merged**
